### PR TITLE
PLANET-5292 Add toggle for Engaging Networks

### PR DESCRIPTION
**Pull request by Inwerpsel**
Ref: https://jira.greenpeace.org/browse/PLANET-5292

Because we use CMB2 [we can't set the default to "on"](https://github.com/CMB2/CMB2/issues/809), so instead we need to set this to "on" in a post deploy script. To more easily do this I added a simple implementation of migration scripts. A new WP option is added to keep track of which scripts are already run (identified by FQCN). In `Activator` it will then run all scripts that haven't run yet, just after updating the capabilities.

Since we know whether a script has already run against the db, we can use this as a condition in our code. That allows us to avoid having to split changes across multiple PRs and deploys. It does require an additional cleanup PR to remove that check once it becomes true for everyone.

For this ticket it allows us to only check the feature toggle after the upgrade script has run, otherwise it's considered as "on".

### CODE REVIEW

File: src/Activator.php [hunk 1]
```diff
@@ -27,5 +29,33 @@ protected function hooks() {
...
+		ini_set( 'display_errors', 1 );
+		// Fetch migration ids that have run from WP option.
+
+		$log = MigrationLog::from_wp_options();
+
+		/**
+		 * @var Migration[] $migrations
+		 */
+		$migrations = [
+			M001EnableEnFormFeature::class,
```
Comment (Inwerpsel) [inline src/Activator.php:45]: We might want to replace this with something that automatically loads all files in the `\Migrations` namespace.

**Review by sagarsdeshmukh — APPROVED** [review]
LGTM 💯

File: src/Migration.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	 * @return string The unique identifier.
+	 */
+	public static function get_id(): string {
+		return static::class;
+	}
+
+	/**
+	 * Perform the actual migration.
+	 */
+	abstract public static function run(): void;
```
[the remark this answers is not in the document]
Comment (Inwerpsel) [reply src/Migration.php:21]: Can you give an example of a dependency a migration could have?

I think I made these static because in essence these migrations are just a function we want to run. That function has access to the same globally available things which are accessible everywhere when using WP (i.e. data from the db and settings in wp-config.php). So I'm not sure if there's anything left to inject as a dependency/parameter?

File: src/Activator.php [hunk 3]
```diff
@@ -27,5 +29,33 @@ protected function hooks() {
 	 */
 	public static function run(): void {
 		Campaigner::register_role_and_add_capabilities();
+		self::do_migrations();
```
[the remark this answers is not in the document]
Comment (Inwerpsel) [reply src/Activator.php:32]: I started with a separate object, but then realized that this Activator is actually already performing a migration, as it updates the capabilities in the db. Note that the class already did more than register the theme switch hook. The same function is also run through the command `p4-run-activator`. So yeah, we definitely want to clean this up a bit, I'll see what I can make of it :+1:

File: src/Activator.php [hunk 4]
```diff
@@ -27,5 +29,33 @@ protected function hooks() {
...
+			M001EnableEnFormFeature::class,
+		];
+
+		// Loop migrations and run those that haven't run yet.
+		foreach ( $migrations as $migration ) {
+			if ( $log->already_ran( $migration::get_id() ) ) {
+				continue;
+			}
+
+			$migration::run();
```
[the remark this answers is not in the document]
Comment (Inwerpsel) [reply src/Activator.php:55]: This is what it currently looks like https://app.circleci.com/pipelines/github/greenpeace/planet4-test-titan/33/workflows/073ddde0-cb79-4f07-b473-06fd98b1e9d3/jobs/255

File: src/Migration.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (lithrel) [reply src/Migration.php:21]: I'm writing too much there so **TLDR; static leads to no isolation. No isolation leads to unstructured code. Unstructured code leads to pain.** <sub>It also leads to me ranting, I don't usually do this, sorry about that</sub>

> Can you give an example of a dependency a migration could have?  

I guess the shortcode converter could have been one ? Any tool that is useful to the current code and has to be used by the migration, to create a new state. In a perfect world, the migration in this PR would use a WP_Options dependency to get and update those (but of course it doesn't exist 😭). In a future patch, I could use the SplitTwoColumns class to run the existing method that convert old blocks on the fly, to patch all posts.

> That function has access to the same globally available things which are accessible everywhere when using WP  

Yes, exactly. Because we are using Wordpress, we tend to fall into the same flaws (or what I consider as flaws), and this is what I would love to not end up doing. **What I'm actually saying is, instead of using things the Wordpress way and calling everything from anywhere, let's make our code cleaner and tighter, and for the most part consider that the global state doesn't exist**. We obviously have to use it, but isolating our code will make it more resilient (because less interactions), less prone to global state unplanned changes (because no global), and more explicit on its needs and inputs/outputs (because typehinting all the things). (Also, more testable.)
My "beef" with static in general is that it's very often no better than using `global`, or the global state (and global state is hard to follow and often has no structure), and the global state is already completely overused by Wordpress. Of course if you push far enough, dependency injection is a container in the global state, and in the end all the code is in it, but having access to anything from anywhere leads your code to less structure and random undeclared coupling.  
 My beef with static methods here is that because it is static by default:
- no instantiation means you won't even think about passing an object as a function parameter, so you won't typehint what your migration runner actually runs, the scope is not obvious, the mental model is wide, I don't know who's using what and in what context.
- using static methods and global state doesn't invite you to declare your dependencies explicitly; the code is then less obvious, what the class and the methods are using is unknown until I read the full code.  
The smallest and more explicit the scope is, the easiest it is for me to reason about it. It's like reading a recipe with a list of all ingredients and tools at the beginning, I can almost get the taste of it without reading the rest, and I know right away what I'll need.
- testing code coupled to wordpress code means including all of wordpress, I'd love our code to be at least a bit more unit testable. Having a `WP_Options` object to inject or add as a parameter would allow us to test the migration. Having MigrationLog and MigrationList as parameters (as interfaces or mockable object) would allow us to easily test a migration run. Having a specific migration runner class with a specific command to run would allow developers to easily test their migration in isolation during development. I'm not even talking about writing tests, but thinking about making it testable usually leads to a cleaner code on the long term (at least for me).

File: src/MigrationLog.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+		return false;
+	}
+
+	/**
+	 * Record a migration in the log.
+	 *
+	 * @param string $migration_id Id of the migration to log.
+	 */
+	public function add( string $migration_id ): void {
```
[the remark this answers is not in the document]
Comment (Inwerpsel) [reply src/MigrationLog.php:56]: We could achieve more separation of concerns if we create a dedicated `MigrationRecord` object that holds just the data that we want to persist when running the script, without also having to hold the logic of the script itself. Then we can provide that record as a parameter to the script. That way the classes holding the actual migration logic will be as simple as possible, i.e. they hold the function that will be run and nothing else, except for the record, but it needs that to record stuff. Tried that out here https://github.com/greenpeace/planet4-master-theme/pull/1172/files#diff-77810f38e224a79d69eaf60f418eb92eR17

[the remark this answers is not in the document]
Comment (Inwerpsel) [reply src/MigrationLog.php:56]: In the linked PR I didn't intentionally didn't change the signature of `already_ran`, since that string alone is sufficient to do that lookup. Otherwise `already_ran` would be tricky to use like it's used here https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/357/files#diff-e0afc9a2c1803d93ce60bc43e7b3d5faR232

File: src/Migration.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (Inwerpsel) [reply src/Migration.php:21]: These are all valid remarks, and if you have control over the entire code base then there indeed should be no global state. However we cannot change the fact that WP puts certain things in the global scope. At best you can add a layer around these things to access them through an object, but that would only make them seem like they're encapsulated, while in the end any piece of code (plugins) can and will still interact with the values through the global that's being wrapped. Because of that those objects should be treated in the same way as though they were the global they're wrapping, so you're missing out on the main benefit of encapsulation while the code has become significantly more complex.

It's also a bit of a rare case here because of the goal of this class:

Specifically in this case the `Migration` class is intended to facilitate writing a script that will interact with the database. In WP interacting with the db can be done in multiple ways, and for convenience here it makes sense to allow them all, so that you can choose whether to use WP's API (`get_posts`, `update_post_meta`, ...) or write a raw SQL query, whichever makes more sense for the tasks the script needs to do. Actually it's not possible to disallow any of these functions anyway as they're globally available. It's a pity but that's just the API WP exposes.

So because we're dealing with scripts that already depend on using global functions to interact with the db, and so the contents of any migration script will already vary a lot, it helps to put the bare minimum inside of this class, and handle complexity in dedicated separate objects.

So, long story short, while in a lot of cases static methods provide less encapsulation, in this case it actually provides more, as we can provide separate objects and pass them as a parameter to the function (like [here](https://github.com/greenpeace/planet4-master-theme/pull/1172/files#diff-da4f24715e30c74e0a897d4de165372bR41)).

File: src/Activator.php (hunk 1 above)
Comment (Inwerpsel) [reply src/Activator.php]: Just mentioning Flo's branch here too which does exactly that :100: https://github.com/greenpeace/planet4-master-theme/commit/aefab111da63d2a25e0525f1576580baeb95746a

File: src/Activator.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (Inwerpsel) [reply src/Activator.php]: I created a `Migrator` class instead to hold this logic. Pretty much the same as the `MigrationRunner` from your branch, but since it's a class with no members and a single function I made it a static function. We can still change it to an instance method and create `Migrator` instances if we would need that instance to hold state or dependencies, just always doing that in advance will lead to a lot of unused instances hanging around. Basically if you find a class being used this way 
``` ( new MyClass())->doFoo() ```, it de facto does not need the instance.

[the remark this answers is not in the document]
Comment (lithrel) [reply src/Activator.php]: > Basically if you find a class being used this way `( new MyClass())->doFoo()`, it de facto does not need the instance.

We will disagree on this for many years to come 😁 but it doesn't matter too much here 👍 ☮️

[the remark this answers is not in the document]
Comment (Inwerpsel) [reply src/Activator.php]: I mean this in a more neutral way :peace_symbol: : in that `de facto` there will be an instance which is created and completely not used. There is no input in the constructor so any members that instance would have would be static values (unless of course the constructor is pulling things from the global scope, which should definitely not happen in the constructor).

File: src/MigrationScript.php [hunk 6]
```diff
@@ -35,19 +35,16 @@ public static function get_id(): string {
...
 		try {
 			static::execute( $record );
+			$record->was_success();
 		} catch ( Exception $exception ) {
-			throw new MigrationFailed(
-				'Migration ' . $record->get_migration_id() . ' failed. Message: ' . $exception->getMessage(),
-				null,
-				$exception
-			);
+			$message = 'Migration ' . $record->get_migration_id() . ' failed. Message: ' . $exception->getMessage();
```
Comment (Inwerpsel) [inline src/MigrationScript.php:44]: We probably don't want one failing script to block all the rest, so I removed throwing the exception and added the message as a log entry instead.
